lib/src/impl_js/impl_js.random.dart

```diff
@@ -16,8 +16,34 @@ part of 'impl_js.dart';
 
 void fillRandomBytes(TypedData destination) {
   try {
-    subtle.getRandomValues(destination);
+    if (destination is Uint8List) {
+      final newValues = subtle.getRandomValues(destination);
+      destination.setAll(0, newValues as Uint8List);
```

Comment:
What is this good for? `getRandomValues` writes to the first argument it is given, does it not?

Reply:
It seems that dart2js overwrites the value, but dart2wasm does not. Could you please see that the `getRandomValues` test fails once as I will submit a PR later that runs with the `--wasm` option in ci?

Reply:
doesn't that hint at a more fundamental issue.

If you read the documentation for `getRandomValues` it's very clear that it fills the first argument with random bytes:

https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues

It even says it's updated in-place:
![image](https://github.com/user-attachments/assets/cde96e06-890a-4507-84db-ee20ac39405a)


----------

If that's not working when compiling to wasm, then what is happening?

I'd suggest trying to make a minimal example. Because this shouldn't be happening, if we can't pass typedarrays correctly to JS from inside wasm, then something is very wrong.

Reply:
I tried it and the following code works.
In the case of dart2js, the `array.toJS` reflected the value in the `array`, but not in the case of dart2wasm.

```dart
final values = array.toJS;
window.crypto.getRandomValues(values);
return values.toDart;
```

Reply:
Can we make a gist that contains the smallest possible test case where:
```
$ dart test -p chrome -c dart2wasm
```
and
```
$ dart test -p chrome -c dart2js
```
behaves differently.

Then let's report a bug against the SDK and see if we can find a workaround.

--------

If there really is a thing where modifications to a `Uint8List` passed to Javascript aren't reflected back in Dart that's pretty bad. And we'd need to review the entire code base for all places where we pass something to Javascript to check that we don't rely on side effects.

Reply:
https://github.com/dart-lang/sdk/issues/59651

Reply:
It was shared that this behavior is to be expected as per the [dart doc](https://api.dart.dev/dart-js_interop/ListToJSArray/toJS.html) and [article's warning](https://dart.dev/interop/js-interop/js-types). For this reason, it would be a good idea to add a process to overwrite the list on the dart side.

Reply:
Ah, I didn't know that `.toJS` could return a clone.

We should probably review all places in the code where we do a `.toJS` call.

Reply:
hmm, I guess we could work around this as follows:
```dart
import 'dart:js_interop';
import 'dart:typed_data';

import 'package:web/web.dart' as web;
import 'package:test/test.dart';

void main() {
  test('calculate', () {
    final array = Uint8List(16);
    expect(
      array.any((e) => e == 0),
      isTrue,
    );

    final jsArray = array.toJS;
    web.window.crypto.getRandomValues(jsArray);

    // .toJS may cause a clone, so we can't know if
    // jsArray.toDart is the same as the original array.
    final dartArray = jsArray.toDart;
    if (dartArray == array) {
      print('We do not need to do anything here');
    } else {
      array.setAll(0, dartArray);
    }

    expect(
      array.any((e) => e != 0),
      isTrue,
    );
  });
}
```

probably we can pick some better variable names. But there is no need to always do `.setAll`.

Reply:
Oops, calling `import 'package:flutter/foundation.dart' show kIsWasm;` on `crypto_subtle.dart` causes an error.

Reply:
We can just use `bool.fromEnvironment('dart.tool.dart2wasm')` it's the same :rofl: